app/controllers/application_controller.rb

```diff
@@ -5,6 +5,7 @@ class ApplicationController < ActionController::Base
   before_filter :authenticate_user!
   before_filter :load_news
   before_filter :set_episode
+  protect_from_forgery with: :null_session, if: Proc.new { |c| c.request.format == 'application/json' }
```

Comment:
I fail to understand how this would *fix* the problem. It would just [skip protection](http://api.rubyonrails.org/classes/ActionController/RequestForgeryProtection/ClassMethods.html) for `POST`ing json, which we don't do, or do I miss something?

Reply:
The token provided by the rails in the beginning and the time we click submit are different because when ajax calls the markdown controller new auth token is provided so I skipped protection only if request format is of application/json.

Reply:
> The token provided by the rails in the beginning and the time we click submit are different because when ajax calls the markdown controller new auth token is provided

Sure thing, I understand the problem :-) 

> so I skipped protection only if request format is of application/json.

What I don't understand is the solution.

1. We **GET** `/markdown/preview?source='## blah'` as **json**. ` protect_from_forgery` isn't used for **GET**, hence your change doesn't apply.
2. Then we **POST** the same data to `/projects` as **html**.  Your `protect_from_forgery` `before_filter` isn't used, as it's not **json**, hence your change doesn't apply.

How does that **fix** the problem?

Reply:
my apologies for the late reply, I used
> protect_from_forgery with: :null_session, if: Proc.new { |c| c.request.format == 'application/json' }

after checking these links : [1](https://coderwall.com/p/8z7z3a/rails-4-solution-for-can-t-verify-csrf-token-authenticity-json-requests) & [2](https://stackoverflow.com/a/10049965/5129516) ,since i dont want to close csrf check.
___
> protect_from_forgery isn't used for GET, hence your change doesn't apply.

^ I didnt get any strong proof but I think maybe in line:8 protect_from_forgery is triggered only when request.format is application/json 🤔

Reply:
Skipping forgery protection in general is not something we want to do right. But as I said, the only type of requests where forgery protection is running are POST requests. I really don't think you can disable the token update for the session like this by skipping this.

Can you try something? When rendering the preview also change the token of the input field? 

```html
<input name="authenticity_token" value="$TOKEN" type="hidden">
```

There is a view helper called `form_authenticity_token` that gives you the current token.